### PR TITLE
Programmable maidroid core

### PR DESCRIPTION
**Pull request by HybridDog**
This makes it possible to load programs written into default written book, the droid interprets and executes the code stored in the first book with title "main" in its inventory.
It depends on that mod:
https://github.com/hybriddog/pdisc
The maidroid instruction set needs implementation, currently it can only beep and set yaw.
And the usage of this core needs (maybe ingame) documentation.

### CODE REVIEW

File: maidroid_core/cores/custom.lua [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	add timediff,beep
+
+	neg timediff
+	add timediff,1000000
+	usleep timediff; should continue 1s after previous beep
+jmp loop_start
+]]
+
+local function on_start(self, metastring)
+	local parsed_code = pdisc.parse(metastring or dummycode)
```
Comment (tacigar) [inline maidroid_core/cores/custom.lua]: In my env, when I put the custom core to the core slot of maidroid, metastring is empty string.
The following code goes well.

```lua
local parsed_code
if metastring ~= nil and metastring ~= "" then
	parsed_code = pdisc.parse(metastring)
else
	parsed_code = pdisc.parse(dummycode)
end
```

**Comment by tacigar** [comment]
Why did you rename `modpack.txt` to `init.lua` ?
In my env, it doesn't work.

File: maidroid_tool/api.lua [hunk 2]
```diff
@@ -52,13 +52,23 @@ function maidroid_tool.register_writer(nodename, options)
...
+				meta:set_float("time", time + elapsed)
 				meta:set_string("formspec", formspec.active(time))
 			else
 				meta:set_float("time", -1)
 				meta:set_string("output", "")
 				meta:set_string("formspec", formspec.inactive)
-				inventory:set_stack("main", 1, ItemStack(output))
+				local resultitem = ItemStack(output)
+
+				if output == "maidroid_core:custom" then
```
Comment (tacigar) [inline maidroid_tool/api.lua]: hmm...
I don't want to embed the specialized code for the custom core into the core_writer.
I would rather create a custom core the same way as other cores, and create a new node that write a program to the custom core.

**Comment by tacigar** [comment]
Your idea is great!
I learned a lot. Thanks!

File: maidroid_tool/api.lua (hunk 2 above)
Comment (HybridDog) [reply maidroid_tool/api.lua]: Yes, this is makeshift, l used it to test if the pdisc mod works.
The custom core also needs to be writeable, else players have to make a new core every time they change/test/adjust their program.

File: maidroid_core/cores/custom.lua (hunk 1 above)
Comment (HybridDog) [reply maidroid_core/cores/custom.lua]: thanks

**Comment by HybridDog** [comment]
> Why did you rename modpack.txt to init.lua ?

That was part of a workaround for #136.

**Comment by ghost** [comment]
How about loading a program from the book which in their inventory?

**Comment by HybridDog** [comment]
Nice idea, an OCR core could be added which makes the maidroid use the book (maybe with the title "main").

**Comment by tacigar** [comment]
> That was part of a workaround for #136.
> l rebased the branch now.

Oh, sorry.

**Comment by HybridDog** [comment]
If the maidroid loots a public chest of some other player, that player could add a book with a maidroid program, when the droid takes the book and some time later restarts its program, it could execute that book, i.e. it got hacked.

I added changes and rebased the PR, now code is read from the book, the programmable core got a texture and requires white dye for crafting(writing).

File: maidroid_core/cores/ocr.lua [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	self.thread:continue()
+end
+
+local function on_pause(self)
+	self.thread:flush()
+end
+
+local function on_stop(self)
+	self.thread:exit()
+
```
Comment (tacigar) [inline maidroid_core/cores/ocr.lua:153]: Please remove `self.thread` by `self.thread = nil`.

**Comment by tacigar** [comment]
> If the maidroid loots a public chest of some other player, that player could add a book with a maidroid program, when the droid takes the book and some time later restarts its program, it could execute that book, i.e. it got hacked.

I will add an ownership system.
It will permit only the owner of the maidroid to put items into the maidroid's inventory.

But, maidroids pickup all dropped items...
If they pickup a dropped written book and restarts, they might execute that book... 😢

**Comment by tacigar** [comment]
> I added changes and rebased the PR, now code is read from the book, the programmable core got a texture and requires white dye for crafting(writing).

Great !!

File: maidroid_core/cores/ocr.lua (hunk 3 above)
Comment (HybridDog) [reply maidroid_core/cores/ocr.lua:153]: fixed

**Comment by HybridDog** [comment]
> If they pickup a dropped written book and restarts, they might execute that book

The book item also has an owner in its metadata.

**Comment by tacigar** [comment]
Please show me a sample code for the ocr core?

**Comment by HybridDog** [comment]
```
start: sleep 1
beep
jump 0.9
jmp start

**Comment by tacigar** [comment]
Good!
I'll merge the pull request!!

**Comment by HybridDog** [comment]
l added the dig instruction, but when the maidroid object digs something, it doesn't get the drops (tested in creative mode).
https://github.com/HybridDog/maidroid/blob/pdisc_core/maidroid_core/cores/ocr.lua#L102

**Comment by tacigar** [comment]
Thanks!
Would you show me an example code using dig instruction?

**Comment by HybridDog** [comment]
```
getpos p.x,p.y,p.z; get maidroid position

start:
add p.y,-1; get node below the previous one
dig $p,success,msg_fail
jmp start,success

; should print failure message, somehow doesnt work
mov msg,$digging failed: 
tostring msg_fail
add msg,msg_fail
print msg
flush

sleep 1; wait a bit after digging failed
jmp start

**Pull request merged**